### PR TITLE
Check base python against env name and raise if differs

### PR DESCRIPTION
**Pull request by gaborbernat**
Resolves https://github.com/tox-dev/tox/issues/1840

Allow enforcing the env name via ignore_base_python_conflict config
value.

### CODE REVIEW

File: tests/tox_env/python/test_python_api.py [hunk 1]
```diff
@@ -65,3 +66,61 @@ def test_diff_msg_added_removed_changed() -> None:
...
+    project = tox_project({"tox.ini": ini})
+    result = project.run("c", "-e", f"py{py_ver}", "-k", "base_python")
+    result.assert_success()
+    if ignore_base_python_conflict:
+        out = f"[testenv:py{py_ver}]\nbase_python = py{py_ver}\n"
+    else:
+        coma_in_exc = sys.version_info[0:2] <= (3, 6)
+        out = (
+            f"[testenv:py{py_ver}]\nbase_python = # Exception: Fail('env name py{py_ver} conflicting with"
+            f" base python py{py_ver_next}'{',' if coma_in_exc else ''})\n"
```
Comment (jugmac00) [inline tests/tox_env/python/test_python_api.py:124]: Should this be `comma_in_exc`?

File: tests/tox_env/python/test_python_api.py [hunk 2]
```diff
@@ -65,3 +66,61 @@ def test_diff_msg_added_removed_changed() -> None:
...
+        if ignore_base_python_conflict:
+            result = Python._validate_base_python(env_name, base_python, ignore_base_python_conflict)
+            assert result == [env_name]
+        else:
+            msg = f"env name {env_name} conflicting with base python {conflict[0]}"
+            with pytest.raises(Fail, match=msg):
+                Python._validate_base_python(env_name, base_python, ignore_base_python_conflict)
+    else:
+        result = Python._validate_base_python(env_name, base_python, ignore_base_python_conflict)
+        assert result is base_python
```
Comment (jugmac00) [inline tests/tox_env/python/test_python_api.py:105]: I am not a big fan of using `if` and `else` in tests and I would not mind to write three small tests, which would duplicate a couple of lines, but then would be easier to reason about - but that is possible a personal preference :-)

**Review by jugmac00 — CHANGES_REQUESTED** [review]
With a `tox.ini` as the following...

```
[tox]
envlist =
    py36,
    py37,
    py38,
    py39,
    pre-commit,
    mypy,
    coverage,
    docs,
ignore_basepython_conflict = true

[testenv]
basepython = python3.9
description = run the tests with pytest
extras = test
deps = pdbpp
commands = pytest {posargs}

[testenv:py36]
commands = python -c "print('hello')"
```

... there are errors no matter whether `ignore_basepython_conflict` is set to `true` or `false`.

```
❯ tox4dev -e py36 -r  # ignore... = false
py36: remove tox env folder /home/jugmac00/Projects/flask-reuploaded/.tox/4/py36
.pkg: remove tox env folder /home/jugmac00/Projects/flask-reuploaded/.tox/4/.pkg
py36: failed with env name py36 conflicting with base python python3.9
  py36: FAIL code 1 (0.00 seconds)
  evaluation failed :( (0.03 seconds)

❯ tox4dev -e py36 -r  # ignore... = true
py36: remove tox env folder /home/jugmac00/Projects/flask-reuploaded/.tox/4/py36
.pkg: remove tox env folder /home/jugmac00/Projects/flask-reuploaded/.tox/4/.pkg
py36: failed with env name py36 conflicting with base python python3.9
  py36: FAIL code 1 (0.00 seconds)
  evaluation failed :( (0.03 seconds)
```

While the `tox3` docs state this will become an error in a future version independent of the setting, this does not match the documentation in this pull request. Also, I remember you said we try to stay compatible with `tox3` as much as possible.

File: docs/config.rst [hunk 3]
```diff
@@ -448,10 +448,9 @@ Python
     do so. To help with sanity of users a error will be raised whenever the environment name version does not matches
     up with this expectation.
 
-    Furthermore, we allow hard enforcing this rule (and bypassing the warning) by setting this flag to ``true``. In such
-    cases we ignore the :ref:`base_python` and instead always use the base python implied from the Python name. This
-    allows you to configure :ref:`base_python` in the :ref:`base` section without affecting environments that have
-    implied base python versions.
+    Furthermore, we allow hard enforcing this rule by setting this flag to ``true``. In such cases we ignore the
+    :ref:`base_python` and instead always use the base python implied from the Python name. This allows you to configure
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply docs/config.rst:452]: Sure, capital all the way.

File: tests/tox_env/python/test_python_api.py (hunk 2 above)
Comment (gaborbernat) [reply tests/tox_env/python/test_python_api.py:105]: IMHO It's a balancing act :blush: made some changes, but did not take it all the way.

**Comment by gaborbernat** [comment]
> I remember you said we try to stay compatible with `tox3` as much as possible.

With some exceptions, this is one of those. And you're right, I interpreted this config as env specific, but is global in tox 3 (hence why wasn't working with your example). Now should be ok.

File: tests/tox_env/python/test_python_api.py (hunk 2 above)
Comment (jugmac00) [reply tests/tox_env/python/test_python_api.py]: Thank you!

**Review by jugmac00 — COMMENTED** [review]
Works like a charm!

**Pull request merged**